### ISSUE TITLE
Rest API customer permissions

### ISSUE THREAD

**Issue by maciekpaprocki** [woocommerce/woocommerce#17983]
It seems like there's a problem with permissions in rest api. 

By default products (and maybe other endpoints) reading should be open to public and show only visible products. 

### Reproduce error

- Go to wc/v2/products endpoint in rest api as non logged user. You will see you are not able to list product. 
- Go to wc/v2/products as customer in rest api. You still see not enough permissions to list products. 

### Expected behaviour

Products should be open for both non logged as well as user logged in as customers

### Code solution

It seems like `woocommerce_rest_check_permissions()` is checking if user can read_private_posts instead of read/read_posts (or just making it true if `$post_type->has_archive` in archive). 

### Temporary solution

This completely not tested and I am not sure what are exact implications, when it comes to security, but basically, this seems to open listing for logged in users: 

```
add_filter( 'woocommerce_rest_check_permissions',
	function ( $permission, $context, $object_id, $post_type ) {
		if ( $context !== 'read' ) {
			return $permission;
		}

		$post_type_object = get_post_type_object( $post_type );

		if ( $object_id ) {
			return current_user_can( $post_type_object->cap->read_post, $object_id );
		}

		return $post_type_object->has_archive;

	},
	10,
	4
);

```
To open it for all users (not logged in too) it would be probably something like this:

```
add_filter( 'woocommerce_rest_check_permissions',
	function ( $permission, $context, $object_id, $post_type ) {

		$post_type_object = get_post_type_object( $post_type );
		return $post_type_object->has_archive;
	},
	10,
	4
);
```

**Comment by claudiosanches** [comment]
This is not an error or issue, but default we don't exposure any endpoint, since they may contain sensitive informations, like downlodable products.
If you want to do it, go ahead and use our filters to update the permissions for your application, but we can't exposure all by default.

**Comment by maciekpaprocki** [comment]
Hmm. The answers should have filtered informations or woocommerce should add custom fields to default wp/v2 product endpoint then. That's how wp rest api works.

It's also against `wp capabilities` which was designed to work in the situation exactly like that. 

It would be probably good to mention that in docs. 

At this moment creating an app for clients, not managers require you to basically write your custom rest endpoints (as assuming that other developers think that API is not public they might extend it with sensitive info). That's massive disadvantage over other commerce systems like Magento, Sylius and basically all the other e-commerce software.

**Comment by claudiosanches** [comment]
@maciekpaprocki we mention that all endpoints require authentication.

**Comment by maciekpaprocki** [comment]
@claudiosanches yes, but as long as you don't authenticate with admin/shop_manager you will still not get data. In wp rest api like for example basic auth, you can authenticate as subscriber for example, which doesn't give you enough permissions. 

Maybe adding: "To use API endpoints you need to authorize that you can manage post type" 

By the way, except of this small issue with authorization I am really happy with Rest API, it's as far the easiest api I ever worked on, with great documentation and support. Well done!

**Comment by claudiosanches** [comment]
@maciekpaprocki feel free to submit a Pull Request in our REST API docs repository: https://github.com/woocommerce/woocommerce-rest-api-docs

**Comment by claudiosanches** [comment]
Finally I have more time now to explain.

So, only admins and shop managers have permissions to create REST API keys too, both in admin screen and under our Application Authentication Endpoint: https://woocommerce.github.io/woocommerce-rest-api-docs/#authentication
We always check for `current_user_can( 'manage_woocommerce' )`, that's why only those kind of user levels can create REST API keys.

As I said before, all endpoints contains a lot of sensitive data that should not be accessed by users.
Creating an app should be used using an admin or shop manager keys, since you will need in all cases a high level user keys in order to create orders or access coupons, if you are trying to create a app to only access products in an store and take orders.

**Comment by maciekpaprocki** [comment]
Hi @claudiosanches ,

Thank you very much for your swift answers. Basically, I only now realised that wc rest api is shop management api not shop api. I understand that now, however, that's a very confusing and makes API unusable for development of for example ios app (which will not be only for managers).

Let's take look at standard wordpress api v2. Users are allowed to read all the public content by default (why wouldn't they be), but to for example update posts they need to be not only authenticated but also authorized by checking do they have right permissions to edit or create entities. This same goes to reading for example drafts, where user which is not authorized will not be allowed to see them. The whole idea behind it is to implement full backend and frontend of wp. Woo does implement backend part, but does not implement front end. 

There's no point in giving app client manager privileges. In fact, that can cause a lot of issues. They should be only able to access what they need. In case someone reverse engineer the API, they will get access to everything manager have, which doesn't seem like a good idea. 

I understand that's basically request for a feature at this point, not a bug with permissions, but I believe it would be great if we could actually have fully working authorization with rest api, this same as default wordpress does it.

**Comment by claudiosanches** [comment]
@maciekpaprocki I understand your points, but our REST API has been created before WordPress REST API, not to mention that we are dealing with e-commerce data, more sensitive than just posts. People have made this decision and have worked with it creating apps without any trouble, there is a lot of extensions today not only to manage shops.
Of course, if you are trying to use as an authentication/login flow you will need write some code to filter and allow it, by default is not possible, but we still allow it.

**Comment by kidswong999** [comment]
@maciekpaprocki 
Hi, I use your code, paste in the `woocommerce/includes/api/class-wc-rest-authentication.php` at `__construct` function.

And I get `/wp-json/wc/v2/products` and `/wp-json/wc/v2/products/id` is OK.

But, get `/wp-json/wc/v2/products/id/variations` or `/wp-json/wc/v2/products/categories` are Forbidden.
I don't know why, they all use `get_items_permissions_check`.
😊
Thanks!

**Comment by kidswong999** [comment]
product has `$post_type_object->has_archive`, so ... 

Are there some solution to classify permissions?

**Comment by hendrik244** [comment]
Hi @maciekpaprocki,

I just ran into a similar problem as you did.
I have set up a webshop, where you can buy license subscriptions for my app.
As I just followed the Woo API Documentation ("allow users to generate API keys for your APP") I thought, that it would be possible for me to use the Application Authentication Endpoint to let the user generate API Keys for hisself and I get his licenses he has subscripted to, this way. 
When I first tried it with my user (admin) it worked perfectly. Now I tried it with a "normal" user and ran into the permission problem.
How did you solved your problem? Did you customized the API to get it worked for you? Can you provide your solution?

Best regards and thank you,
Hendrik

**Comment by aristech** [comment]
I have the same problem guys! I get admins to manage the shop but no users to even view the products.
I'm sure i can get the products with custom endpoints but having woocomerce api and not using it for basic functionality is pointless!

**Comment by rrennick** [comment]
> I get admins to manage the shop but no users to even view the products.

This issue was related to creating API keys vs using the API. If you create a set of READ api keys, those keys can be used by any API client to retrieve data from the API.

**Comment by JoshMoreno** [comment]
I think the problem a lot of people are having is that read access doesn't always cover all use cases. For example, if I want to use vue or react for my front end and wp + woocommerce in a headless fashion, I'm only limited to GET and OPTION requests when leveraging woocommerce's api.

**woocommerce/includes/class-wc-rest-authentication.php around line 64**
```
// Allow third party plugins use our authentication methods.
$third_party = ( false !== strpos( $request_uri, $rest_prefix . 'wc-' ) );

return apply_filters( 'woocommerce_rest_is_request_to_rest_api', $woocommerce || $third_party );
```

A use case for post requests would be adding to cart using a plugin like this - https://wordpress.org/plugins/cart-rest-api-for-woocommerce/

So now, I have to grant read and write access. Okay, but there's still capabilities WooCommerce defines; Great! But, if I grant the api user access to read_products, they can also read unpublished products! Also, we have a key and a secret which I've only ever seen where one or both are meant to be used server side and not exposed on the client side. I'm very hesitant to trust exposing both keys client side and trust that I've covered all possible vulnerabilities. 

The alternative would be to create a proxy or custom endpoints to just grab the data myself. But that defeats the purpose of the REST API.

It's just a lot of work and does seem very geared towards shop management instead of just being a complete solution. I hope this changes in the future.

**Comment by claudiosanches** [comment]
@JoshMoreno in fact you should use `woocommerce_rest_check_permissions` filter, so you can manipulate each required capability.

**Comment by sunxiyuan** [comment]
I am also troubled by this.
Currently I have to rewrite a plug-in to provide a new API to serve users. This API inherits from Woo's API and passes the value to Woo after filtering the parameters.
This is troublesome, it would be great if Woo could provide Rest Api similar to WordPress

**Comment by davidhealey** [comment]
A user should at least have access to their orders, the things they can see when they log in to the website...

**Comment by thd-fox** [comment]
@claudiosan I found this issue right now and this will block anyone else from using woocommerce headless. Why you're not taking this into roadmap? Once there is security by client credentials, that can be assigned to at least for read access without 'manage_woocommerce' rights, and a shop owner puts this into a headless client app, there is responsibility at shop owner.

Sad to see you are blocking awesomeness. I bet on woocommerce sinc I started working with it professionally.

Now there must be a third party solution, again :( Or a proprietary solution...
